// @genkit-ai/google-genai/lib/googleai/client.mjs
import { GenkitError } from "genkit";
import { logger } from "genkit/logging";
import { runInNewSpan } from "genkit/tracing";
import {
  buildTraceMetadataInput,
  extractErrMsg,
  getGenkitClientHeader,
  parseRetryAfterMs,
  processStream
} from "../common/utils.js";
const tracingHooks = { runInNewSpan };
async function createInteraction(apiKey, createInteractionRequest, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: "interactions",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    body: JSON.stringify(createInteractionRequest),
    isInteraction: true
  });
  return maybeTraceRequest(url, fetchOptions, {
    request: createInteractionRequest,
    clientOptions
  });
}
async function getInteraction(apiKey, interactionId, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `interactions/${interactionId}`,
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "GET",
    apiKey,
    clientOptions,
    isInteraction: true
  });
  return maybeTraceRequest(url, fetchOptions, {
    clientOptions
  });
}
async function cancelInteraction(apiKey, interactionId, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `interactions/${interactionId}/cancel`,
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    isInteraction: true
  });
  try {
    await makeRequest(url, fetchOptions);
    throw new GenkitError({
      status: "CANCELLED",
      message: "successfully cancelled"
    });
  } catch (e) {
    if (e instanceof GenkitError && e.status === "CANCELLED") {
      return {
        id: interactionId,
        status: "cancelled"
      };
    }
    throw e;
  }
}
async function listModels(apiKey, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: "models",
    queryParams: "pageSize=1000",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "GET",
    apiKey,
    clientOptions
  });
  const response = await makeRequest(url, fetchOptions);
  const modelResponse = JSON.parse(await response.text());
  return modelResponse.models;
}
async function generateContent(apiKey, model, generateContentRequest, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `models/${model}`,
    resourceMethod: "generateContent",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    body: JSON.stringify(generateContentRequest)
  });
  return maybeTraceRequest(url, fetchOptions, {
    model,
    request: generateContentRequest,
    clientOptions
  });
}
async function generateContentStream(apiKey, model, generateContentRequest, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `models/${model}`,
    resourceMethod: "streamGenerateContent",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    body: JSON.stringify(generateContentRequest)
  });
  return maybeTraceRequest(url, fetchOptions, {
    model,
    request: generateContentRequest,
    streaming: true,
    clientOptions,
    processFn: async (response) => {
      return processStream(response);
    }
  });
}
async function embedContent(apiKey, model, embedContentRequest, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `models/${model}`,
    resourceMethod: "embedContent",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    body: JSON.stringify(embedContentRequest)
  });
  return maybeTraceRequest(url, fetchOptions, {
    model,
    request: embedContentRequest,
    clientOptions
  });
}
async function imagenPredict(apiKey, model, imagenPredictRequest, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `models/${model}`,
    resourceMethod: "predict",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    body: JSON.stringify(imagenPredictRequest)
  });
  return maybeTraceRequest(url, fetchOptions, {
    model,
    request: imagenPredictRequest,
    clientOptions
  });
}
async function veoPredict(apiKey, model, veoPredictRequest, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: `models/${model}`,
    resourceMethod: "predictLongRunning",
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "POST",
    apiKey,
    clientOptions,
    body: JSON.stringify(veoPredictRequest)
  });
  return maybeTraceRequest(url, fetchOptions, {
    model,
    request: veoPredictRequest,
    clientOptions
  });
}
async function veoCheckOperation(apiKey, operation, clientOptions) {
  const url = getGoogleAIUrl({
    resourcePath: operation,
    clientOptions
  });
  const fetchOptions = getFetchOptions({
    method: "GET",
    apiKey,
    clientOptions
  });
  return maybeTraceRequest(url, fetchOptions, {
    clientOptions
  });
}
function getGoogleAIUrl(params) {
  const DEFAULT_API_VERSION = "v1beta";
  const DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com";
  const apiVersion = params.clientOptions?.apiVersion || DEFAULT_API_VERSION;
  const baseUrl = params.clientOptions?.baseUrl || DEFAULT_BASE_URL;
  let url = `${baseUrl}/${apiVersion}/${params.resourcePath}`;
  if (params.resourceMethod) {
    url += `:${params.resourceMethod}`;
  }
  if (params.queryParams) {
    url += `?${params.queryParams}`;
  }
  if (params.resourceMethod === "streamGenerateContent") {
    url += `${params.queryParams ? "&" : "?"}alt=sse`;
  }
  return url;
}
function getFetchOptions(params) {
  const fetchOptions = {
    method: params.method,
    headers: getHeaders(
      params.apiKey,
      params.clientOptions,
      params.isInteraction
    )
  };
  if (params.body) {
    fetchOptions.body = params.body;
  }
  const signal = getAbortSignal(params.clientOptions);
  if (signal) {
    fetchOptions.signal = signal;
  }
  return fetchOptions;
}
function getAbortSignal(clientOptions) {
  const hasTimeout = (clientOptions?.timeout ?? -1) >= 0;
  if (clientOptions?.signal !== void 0 || hasTimeout) {
    const controller = new AbortController();
    if (hasTimeout) {
      setTimeout(() => controller.abort(), clientOptions?.timeout);
    }
    if (clientOptions?.signal) {
      clientOptions.signal.addEventListener("abort", () => {
        controller.abort();
      });
    }
    return controller.signal;
  }
  return void 0;
}
function getHeaders(apiKey, clientOptions, isInteraction) {
  let customHeaders = {};
  if (clientOptions?.customHeaders) {
    customHeaders = structuredClone(clientOptions.customHeaders);
    delete customHeaders["x-goog-api-key"];
    delete customHeaders["x-goog-api-client"];
  }
  const headers = {
    ...customHeaders,
    "Content-Type": "application/json",
    "x-goog-api-client": getGenkitClientHeader()
  };
  if (isInteraction) {
    headers["Api-Revision"] = "2026-05-20";
  }
  if (apiKey) {
    headers["x-goog-api-key"] = apiKey;
  }
  return headers;
}
async function maybeTraceRequest(url, fetchOptions, traceOptions) {
  const call = async () => {
    const response = await makeRequest(url, fetchOptions);
    let processedResponse;
    if (traceOptions.processFn) {
      processedResponse = traceOptions.processFn(response);
    } else {
      processedResponse = await response.json();
    }
    return processedResponse;
  };
  if (traceOptions.clientOptions?.experimental_debugTraces) {
    return tracingHooks.runInNewSpan(
      { metadata: { name: "httpRequest" } },
      async (metadata) => {
        metadata.input = buildTraceMetadataInput(
          url,
          fetchOptions,
          traceOptions
        );
        const processedResponse = await call();
        if (traceOptions.streaming) {
          metadata.output = "[Streaming Response]";
        } else {
          metadata.output = processedResponse;
        }
        return processedResponse;
      }
    );
  }
  return call();
}
async function makeRequest(url, fetchOptions) {
  try {
    const response = await fetch(url, fetchOptions);
    if (!response.ok) {
      let errorText = await response.text();
      let errorMessage = errorText;
      let errorDetail;
      try {
        const json = JSON.parse(errorText);
        errorDetail = json;
        if (json.error && json.error.message) {
          errorMessage = json.error.message;
          if (Array.isArray(json.error.details)) {
            const detailsText = json.error.details.map((d) => {
              if (d.detail && typeof d.detail === "string") {
                const match = d.detail.match(/\[ORIGINAL ERROR\]\s*([^[]+)/);
                const detailText = match ? match[1].trim() : d.detail;
                return `${detailText}
Raw: ${JSON.stringify(d, null, 2)}`;
              }
              return JSON.stringify(d, null, 2);
            }).filter(Boolean).join("\n");
            if (detailsText) {
              errorMessage += `
Details:
${detailsText}`;
            }
          }
        }
      } catch (e) {
      }
      let status = "UNKNOWN";
      switch (response.status) {
        case 429:
          status = "RESOURCE_EXHAUSTED";
          break;
        case 400:
          status = "INVALID_ARGUMENT";
          break;
        case 499:
          status = "CANCELLED";
          break;
        case 500:
          status = "INTERNAL";
          break;
        case 503:
          status = "UNAVAILABLE";
          break;
      }
      const retryAfterHeader = response.headers.get("retry-after");
      const retryAfterMs = retryAfterHeader ? parseRetryAfterMs(retryAfterHeader) : void 0;
      const responseMetadata = retryAfterMs !== void 0 ? { retryAfterMs } : void 0;
      throw new GenkitError({
        status,
        message: `Error fetching from ${url}: [${response.status} ${response.statusText}] ${errorMessage}`,
        detail: errorDetail,
        responseMetadata
      });
    }
    return response;
  } catch (e) {
    logger.error(e);
    if (e instanceof GenkitError) {
      throw e;
    }
    throw new Error(`Failed to fetch from ${url}: ${extractErrMsg(e)}`);
  }
}
const TEST_ONLY = {
  getFetchOptions,
  getAbortSignal,
  getHeaders,
  makeRequest,
  tracingHooks
};
export {
  TEST_ONLY,
  cancelInteraction,
  createInteraction,
  embedContent,
  generateContent,
  generateContentStream,
  getGoogleAIUrl,
  getInteraction,
  imagenPredict,
  listModels,
  tracingHooks,
  veoCheckOperation,
  veoPredict
};
