// @genkit-ai/google-genai/lib/googleai/embedder.mjs
import {
  embedderActionMetadata,
  z
} from "genkit";
import { embedderRef } from "genkit/embedder";
import { embedder as pluginEmbedder } from "genkit/plugin";
import { toGeminiMessage } from "../common/converters.js";
import { isKnownKey } from "../common/utils.js";
import { embedContent } from "./client.js";
import {
  TaskTypeSchema
} from "./types.js";
import {
  calculateApiKey,
  checkApiKey,
  checkModelName,
  extractVersion
} from "./utils.js";
const EmbeddingConfigSchema = z.object({
  /** Override the API key provided at plugin initialization. */
  apiKey: z.string().optional(),
  /**
   * The `task_type` parameter is defined as the intended downstream application to help the model
   * produce better quality embeddings.
   **/
  taskType: TaskTypeSchema.optional(),
  title: z.string().optional(),
  version: z.string().optional(),
  /**
   * The `outputDimensionality` parameter allows you to specify the dimensionality of the embedding output.
   * By default, `gemini-embedding-2` and `gemini-embedding-2-preview` generate embeddings with 3072 dimensions,
   * while `gemini-embedding-001` generates 768 dimensions.
   * By selecting a smaller output dimensionality, users can save memory and storage space, leading to more efficient computations.
   **/
  outputDimensionality: z.number().min(1).optional()
}).passthrough();
function commonRef(name, info, configSchema = EmbeddingConfigSchema) {
  return embedderRef({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      dimensions: 768,
      supports: {
        input: ["text"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("embedder");
const KNOWN_MODELS = {
  "gemini-embedding-2-preview": commonRef("gemini-embedding-2-preview", {
    dimensions: 3072,
    supports: {
      input: ["text", "image", "video"]
    }
  }),
  "gemini-embedding-2": commonRef("gemini-embedding-2", {
    dimensions: 3072,
    supports: {
      input: ["text", "image", "video"]
    }
  }),
  "gemini-embedding-001": commonRef("gemini-embedding-001")
};
function isEmbedderName(value) {
  return value.startsWith("gemini-embedding-");
}
function model(version, config = {}) {
  const name = checkModelName(version);
  if (isKnownKey(name, KNOWN_MODELS)) {
    const known = KNOWN_MODELS[name];
    return embedderRef({
      name: known.name,
      info: known.info,
      configSchema: known.configSchema,
      config
    });
  }
  return embedderRef({
    name: `googleai/${name}`,
    info: { ...GENERIC_MODEL.info },
    configSchema: GENERIC_MODEL.configSchema,
    config
  });
}
function listActions(models) {
  return models.filter((m) => m.supportedGenerationMethods.includes("embedContent")).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return embedderActionMetadata({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map((name) => defineEmbedder(name, options));
}
function defineEmbedder(name, pluginOptions) {
  checkApiKey(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders
  };
  return pluginEmbedder(
    {
      name: ref.name,
      configSchema: ref.configSchema,
      info: ref.info
    },
    async (request, _) => {
      const embedApiKey = calculateApiKey(
        pluginOptions?.apiKey,
        request.options?.apiKey
      );
      const embedVersion = request.options?.version || extractVersion(ref);
      const embeddings = await Promise.all(
        request.input.map(async (doc) => {
          const response = await embedContent(
            embedApiKey,
            embedVersion,
            {
              taskType: request.options?.taskType,
              title: request.options?.title,
              content: toGeminiMessage({ role: "user", content: doc.content }),
              outputDimensionality: request.options?.outputDimensionality
            },
            clientOptions
          );
          const values = response.embedding.values;
          return { embedding: values };
        })
      );
      return { embeddings };
    }
  );
}
const TEST_ONLY = { KNOWN_MODELS };
export {
  EmbeddingConfigSchema,
  TEST_ONLY,
  defineEmbedder,
  isEmbedderName,
  listActions,
  listKnownModels,
  model
};
