// @genkit-ai/google-genai/lib/googleai/index.mjs
import { logger } from "genkit/logging";
import {
  genkitPluginV2
} from "genkit/plugin";
import { extractErrMsg } from "../common/utils.js";
import { listModels } from "./client.js";
import { calculateApiKey } from "./utils.js";
import * as deepResearch from "./deep-research.js";
import * as embedder from "./embedder.js";
import * as gemini from "./gemini.js";
import * as imagen from "./imagen.js";
import * as lyria from "./lyria.js";
import * as veo from "./veo.js";
async function initializer(options) {
  return [
    ...imagen.listKnownModels(options),
    ...gemini.listKnownModels(options),
    ...embedder.listKnownModels(options),
    ...veo.listKnownModels(options),
    ...deepResearch.listKnownModels(options),
    ...lyria.listKnownModels(options)
  ];
}
async function resolver(actionType, actionName, options) {
  switch (actionType) {
    case "model":
      if (veo.isVeoModelName(actionName) || deepResearch.isDeepResearchModelName(actionName)) {
        return void 0;
      } else if (imagen.isImagenModelName(actionName)) {
        return imagen.defineModel(actionName, options);
      } else if (lyria.isLyriaModelName(actionName)) {
        return lyria.defineModel(actionName, options);
      } else {
        return gemini.defineModel(actionName, options);
      }
    case "background-model":
      if (veo.isVeoModelName(actionName)) {
        return veo.defineModel(actionName, options);
      } else if (deepResearch.isDeepResearchModelName(actionName)) {
        return deepResearch.defineModel(actionName, options);
      }
      break;
    // No default
    case "embedder":
      return embedder.defineEmbedder(actionName, options);
  }
  return void 0;
}
async function listActions(options) {
  if (options?.apiKey === false) {
    return [];
  }
  try {
    const apiKey = calculateApiKey(options?.apiKey, void 0);
    const allModels = await listModels(apiKey, {
      baseUrl: options?.baseUrl,
      apiVersion: options?.apiVersion
    });
    return [
      ...gemini.listActions(allModels),
      ...imagen.listActions(allModels),
      ...veo.listActions(allModels),
      ...deepResearch.listActions(allModels),
      ...lyria.listActions(allModels),
      ...embedder.listActions(allModels)
    ];
  } catch (e) {
    logger.error(extractErrMsg(e));
    return [];
  }
}
function googleAIPlugin(options) {
  let listActionsCache;
  return genkitPluginV2({
    name: "googleai",
    init: async () => await initializer(options),
    resolve: async (actionType, actionName) => await resolver(actionType, actionName, options || {}),
    list: async () => {
      if (listActionsCache) return listActionsCache;
      listActionsCache = await listActions(options);
      return listActionsCache;
    }
  });
}
const googleAI = googleAIPlugin;
googleAI.model = (name, config) => {
  if (veo.isVeoModelName(name)) {
    return veo.model(name, config);
  }
  if (deepResearch.isDeepResearchModelName(name)) {
    return deepResearch.model(name, config);
  }
  if (lyria.isLyriaModelName(name)) {
    return lyria.model(name, config);
  }
  if (imagen.isImagenModelName(name)) {
    return imagen.model(name, config);
  }
  return gemini.model(name, config);
};
googleAI.embedder = (name, config) => {
  return embedder.model(name, config);
};
var googleai_default = googleAI;
export {
  googleai_default as default,
  googleAI,
  googleAIPlugin
};
