// @genkit-ai/google-genai/lib/vertexai/lyria.mjs
import {
  modelActionMetadata,
  modelRef,
  z
} from "genkit";
import { model as pluginModel } from "genkit/plugin";
import {
  ensureToolIds,
  fromInteractionSync,
  toInteractionTurn
} from "../common/interaction-converters.js";
import { isKnownKey } from "../common/utils.js";
import { createInteraction, lyriaPredict } from "./client.js";
import { fromLyriaResponse, toLyriaPredictRequest } from "./converters.js";
import {
  calculateRequestOptions,
  checkModelName,
  extractVersion,
  shouldUseLegacyEndpoint
} from "./utils.js";
const LyriaConfigSchema = z.object({
  negativePrompt: z.string().optional().describe(
    "Optional. A description of what to exclude from the generated audio."
  ),
  seed: z.number().optional().describe(
    "Optional. A seed for deterministic generation. If provided, the model will attempt to produce the same audio given the same prompt and other parameters. Cannot be used with sample_count in the same request."
  ),
  sampleCount: z.number().optional().describe(
    "Optional. The number of audio samples to generate. Default is 1 if not specified and seed is not used. Cannot be used with seed in the same request."
  ),
  location: z.string().describe(
    "Lyria is only available in global. If you initialize your plugin with a different region, you must set this to global."
  ).optional()
}).passthrough();
const Lyria3ConfigSchema = z.object({
  location: z.string().describe(
    "Lyria is only available in global. If you initialize your plugin with a different region, you must set this to global."
  ).optional(),
  responseModalities: z.array(z.enum(["TEXT", "IMAGE", "AUDIO"])).describe(
    "The modalities to be used in response. Defaults to AUDIO and TEXT for Lyria."
  ).optional()
}).passthrough();
function isLegacyLyriaRequest(request, name) {
  return shouldUseLegacyEndpoint(name);
}
function isLyria3Request(request, name) {
  return !shouldUseLegacyEndpoint(name);
}
function commonRef(name, info, configSchema = Lyria3ConfigSchema) {
  return modelRef({
    name: `vertexai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        media: true,
        multiturn: false,
        tools: false,
        systemRole: false,
        output: ["text", "media"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("lyria");
const GENERIC_LEGACY_MODEL = commonRef(
  "lyria-002",
  {
    supports: {
      media: true,
      multiturn: false,
      tools: false,
      systemRole: false,
      output: ["media"]
    }
  },
  LyriaConfigSchema
);
const KNOWN_LYRIA_LEGACY_MODELS = {
  "lyria-002": commonRef(
    "lyria-002",
    { ...GENERIC_LEGACY_MODEL.info },
    LyriaConfigSchema
  )
};
const KNOWN_LYRIA_INTERACTIONS_MODELS = {
  "lyria-3-clip-preview": commonRef("lyria-3-clip-preview"),
  "lyria-3-pro-preview": commonRef("lyria-3-pro-preview")
};
const KNOWN_MODELS = {
  ...KNOWN_LYRIA_LEGACY_MODELS,
  ...KNOWN_LYRIA_INTERACTIONS_MODELS
};
function isLyriaModelName(value) {
  return !!value?.startsWith("lyria-");
}
function model(version, config = {}) {
  const name = checkModelName(version);
  if (isKnownKey(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return modelRef({
    name: `vertexai/${name}`,
    config,
    configSchema: Lyria3ConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isLyriaModelName(m.name)).map((m) => {
    const ref = model(m.name);
    return modelActionMetadata({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(clientOptions, pluginOptions) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, clientOptions, pluginOptions)
  );
}
function defineModel(name, clientOptions, pluginOptions) {
  const ref = model(name);
  return pluginModel(
    {
      name: ref.name,
      ...ref.info,
      configSchema: ref.configSchema
    },
    async (request, { abortSignal }) => {
      const clientOpt = calculateRequestOptions(
        { ...clientOptions, signal: abortSignal },
        request.config
      );
      if (isLegacyLyriaRequest(request, name)) {
        const lyriaPredictRequest = toLyriaPredictRequest(request);
        const response = await lyriaPredict(
          extractVersion(ref),
          lyriaPredictRequest,
          clientOpt
        );
        if (!response.predictions || response.predictions.length == 0) {
          throw new Error(
            "Model returned no predictions. Possibly due to content filters."
          );
        }
        return fromLyriaResponse(response, request);
      } else if (isLyria3Request(request, name)) {
        const messages = [...request.messages];
        if (messages.length === 0) throw new Error("No messages provided");
        let responseModalitiesConverted = ["audio", "text"];
        if (request.config?.responseModalities) {
          responseModalitiesConverted = request.config.responseModalities.map(
            (m) => m.toLowerCase()
          );
        }
        const req = {
          model: extractVersion(ref),
          input: ensureToolIds(messages).map(toInteractionTurn),
          response_modalities: responseModalitiesConverted
        };
        const response = await createInteraction(req, clientOpt);
        return fromInteractionSync(response);
      } else {
        throw new Error(`Unsupported model config schema for ${name}`);
      }
    }
  );
}
const TEST_ONLY = {
  GENERIC_MODEL,
  KNOWN_MODELS,
  KNOWN_LYRIA_LEGACY_MODELS,
  KNOWN_LYRIA_INTERACTIONS_MODELS
};
export {
  Lyria3ConfigSchema,
  LyriaConfigSchema,
  TEST_ONLY,
  defineModel,
  isLyriaModelName,
  listActions,
  listKnownModels,
  model
};
