// @genkit-ai/google-genai/lib/vertexai/lyria.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var lyria_exports = {};
__export(lyria_exports, {
  Lyria3ConfigSchema: () => Lyria3ConfigSchema,
  LyriaConfigSchema: () => LyriaConfigSchema,
  TEST_ONLY: () => TEST_ONLY,
  defineModel: () => defineModel,
  isLyriaModelName: () => isLyriaModelName,
  listActions: () => listActions,
  listKnownModels: () => listKnownModels,
  model: () => model
});
module.exports = __toCommonJS(lyria_exports);
var import_genkit = require("genkit");
var import_plugin = require("genkit/plugin");
var import_interaction_converters = require("../common/interaction-converters.js");
var import_utils = require("../common/utils.js");
var import_client = require("./client.js");
var import_converters = require("./converters.js");
var import_utils2 = require("./utils.js");
const LyriaConfigSchema = import_genkit.z.object({
  negativePrompt: import_genkit.z.string().optional().describe(
    "Optional. A description of what to exclude from the generated audio."
  ),
  seed: import_genkit.z.number().optional().describe(
    "Optional. A seed for deterministic generation. If provided, the model will attempt to produce the same audio given the same prompt and other parameters. Cannot be used with sample_count in the same request."
  ),
  sampleCount: import_genkit.z.number().optional().describe(
    "Optional. The number of audio samples to generate. Default is 1 if not specified and seed is not used. Cannot be used with seed in the same request."
  ),
  location: import_genkit.z.string().describe(
    "Lyria is only available in global. If you initialize your plugin with a different region, you must set this to global."
  ).optional()
}).passthrough();
const Lyria3ConfigSchema = import_genkit.z.object({
  location: import_genkit.z.string().describe(
    "Lyria is only available in global. If you initialize your plugin with a different region, you must set this to global."
  ).optional(),
  responseModalities: import_genkit.z.array(import_genkit.z.enum(["TEXT", "IMAGE", "AUDIO"])).describe(
    "The modalities to be used in response. Defaults to AUDIO and TEXT for Lyria."
  ).optional()
}).passthrough();
function isLegacyLyriaRequest(request, name) {
  return (0, import_utils2.shouldUseLegacyEndpoint)(name);
}
function isLyria3Request(request, name) {
  return !(0, import_utils2.shouldUseLegacyEndpoint)(name);
}
function commonRef(name, info, configSchema = Lyria3ConfigSchema) {
  return (0, import_genkit.modelRef)({
    name: `vertexai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        media: true,
        multiturn: false,
        tools: false,
        systemRole: false,
        output: ["text", "media"]
      }
    }
  });
}
const GENERIC_MODEL = commonRef("lyria");
const GENERIC_LEGACY_MODEL = commonRef(
  "lyria-002",
  {
    supports: {
      media: true,
      multiturn: false,
      tools: false,
      systemRole: false,
      output: ["media"]
    }
  },
  LyriaConfigSchema
);
const KNOWN_LYRIA_LEGACY_MODELS = {
  "lyria-002": commonRef(
    "lyria-002",
    { ...GENERIC_LEGACY_MODEL.info },
    LyriaConfigSchema
  )
};
const KNOWN_LYRIA_INTERACTIONS_MODELS = {
  "lyria-3-clip-preview": commonRef("lyria-3-clip-preview"),
  "lyria-3-pro-preview": commonRef("lyria-3-pro-preview")
};
const KNOWN_MODELS = {
  ...KNOWN_LYRIA_LEGACY_MODELS,
  ...KNOWN_LYRIA_INTERACTIONS_MODELS
};
function isLyriaModelName(value) {
  return !!value?.startsWith("lyria-");
}
function model(version, config = {}) {
  const name = (0, import_utils2.checkModelName)(version);
  if ((0, import_utils.isKnownKey)(name, KNOWN_MODELS)) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return (0, import_genkit.modelRef)({
    name: `vertexai/${name}`,
    config,
    configSchema: Lyria3ConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isLyriaModelName(m.name)).map((m) => {
    const ref = model(m.name);
    return (0, import_genkit.modelActionMetadata)({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(clientOptions, pluginOptions) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, clientOptions, pluginOptions)
  );
}
function defineModel(name, clientOptions, pluginOptions) {
  const ref = model(name);
  return (0, import_plugin.model)(
    {
      name: ref.name,
      ...ref.info,
      configSchema: ref.configSchema
    },
    async (request, { abortSignal }) => {
      const clientOpt = (0, import_utils2.calculateRequestOptions)(
        { ...clientOptions, signal: abortSignal },
        request.config
      );
      if (isLegacyLyriaRequest(request, name)) {
        const lyriaPredictRequest = (0, import_converters.toLyriaPredictRequest)(request);
        const response = await (0, import_client.lyriaPredict)(
          (0, import_utils2.extractVersion)(ref),
          lyriaPredictRequest,
          clientOpt
        );
        if (!response.predictions || response.predictions.length == 0) {
          throw new Error(
            "Model returned no predictions. Possibly due to content filters."
          );
        }
        return (0, import_converters.fromLyriaResponse)(response, request);
      } else if (isLyria3Request(request, name)) {
        const messages = [...request.messages];
        if (messages.length === 0) throw new Error("No messages provided");
        let responseModalitiesConverted = ["audio", "text"];
        if (request.config?.responseModalities) {
          responseModalitiesConverted = request.config.responseModalities.map(
            (m) => m.toLowerCase()
          );
        }
        const req = {
          model: (0, import_utils2.extractVersion)(ref),
          input: (0, import_interaction_converters.ensureToolIds)(messages).map(import_interaction_converters.toInteractionTurn),
          response_modalities: responseModalitiesConverted
        };
        const response = await (0, import_client.createInteraction)(req, clientOpt);
        return (0, import_interaction_converters.fromInteractionSync)(response);
      } else {
        throw new Error(`Unsupported model config schema for ${name}`);
      }
    }
  );
}
const TEST_ONLY = {
  GENERIC_MODEL,
  KNOWN_MODELS,
  KNOWN_LYRIA_LEGACY_MODELS,
  KNOWN_LYRIA_INTERACTIONS_MODELS
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Lyria3ConfigSchema,
  LyriaConfigSchema,
  TEST_ONLY,
  defineModel,
  isLyriaModelName,
  listActions,
  listKnownModels,
  model
});
